### PR TITLE
Handle empty application name

### LINKED ISSUE #3349: Empty application name isn't rejected, causes issues
https://github.com/CanonicalLtd/jujucharms.com/issues/500

Steps to reproduce

Add a charm to the canvas (I started with PostgreSQL but verified with K8S)
Configure one of the applications and empty the "Application name"
At this point, I expect a validation error, but it didn't complain... so:

Click deploy changes
Note that nothing happened, but look in your browser's console and see an error - http://paste.ubuntu.com/26003939/

### PR DESCRIPTION
**Pull request by makyo**
Fixes #3349

### CODE REVIEW

**Comment by makyo** [comment]
[Empty application name isn't rejected](https://trello.com/c/FXucWf6D/66-empty-application-name-isnt-rejected)

**Comment by makyo** [comment]
[Empty application name isn't rejected](https://trello.com/c/FXucWf6D/66-empty-application-name-isnt-rejected)

File: jujugui/static/gui/src/app/components/inspector/config/test-config.js [hunk 1]
```diff
@@ -647,6 +647,58 @@ describe('Configuration', function() {
     assert.equal(updateUnit.callCount, 0);
   });
 
+  it('stops setting changes if service name is blank', () => {
+    var option1 = { key: 'option1key', type: 'string' };
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector/config/test-config.js]: can you update these to use `let` and `const` plz.

File: jujugui/static/gui/src/app/components/inspector/config/config.js [hunk 2]
```diff
@@ -140,6 +140,14 @@ class Configuration extends React.Component {
         });
         return;
       }
+      if (!nameValue) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector/config/config.js]: While this solves the immediate problem, why don't we expand on this a bit and test with the same regex as core so we're also sure it's a valid name wrt the allowed characters?

File: jujugui/static/gui/src/app/components/inspector/config/test-config.js (hunk 1 above)
Comment (makyo) [reply jujugui/static/gui/src/app/components/inspector/config/test-config.js]: Crap, sorry. Copypasta.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8289/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8291/

File: jujugui/static/gui/src/app/components/inspector/config/test-config.js [hunk 3]
```diff
@@ -647,6 +647,70 @@ describe('Configuration', function() {
...
+    const domNode = ReactDOM.findDOMNode(component);
+    const name = domNode.querySelector('.string-config-input');
+
+    name.innerText = '';
+    testUtils.Simulate.input(name);
+
+    const save = domNode.querySelector('.button-row--count-2 .button--neutral');
+    testUtils.Simulate.click(save);
+
+    // Make sure it emits a notification if the name exists.
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector/config/test-config.js]: "...if the name **doesn't** exist."?

File: jujugui/static/gui/src/app/components/inspector/config/test-config.js [hunk 4]
```diff
@@ -647,6 +647,70 @@ describe('Configuration', function() {
...
+    assert.equal(service.set.callCount, 0);
+    // Make sure that the unit names weren't updated.
+    assert.equal(updateUnit.callCount, 0);
+
+    name.innerText = '!@#$%';
+    testUtils.Simulate.input(name);
+
+    testUtils.Simulate.click(save);
+
+    // Make sure it emits a notification if the name exists.
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector/config/test-config.js]: "... if the name **is invalid**."?

File: jujugui/static/gui/src/app/components/inspector/config/config.js [hunk 5]
```diff
@@ -140,6 +140,14 @@ class Configuration extends React.Component {
         });
         return;
       }
+      if (!window.jujulib.NAME_PATTERN.test(nameValue)) {
+        props.addNotification({
+          title: 'Invalid application name',
+          message: 'Invalid application name`.',
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/inspector/config/config.js]: There's a ` hanging out there on the end for some reason.

File: jujugui/static/gui/src/app/components/inspector/config/test-config.js (hunk 3 above)
Comment (frankban) [reply jujugui/static/gui/src/app/components/inspector/config/test-config.js]: if the name is empty? ;-)

File: jujugui/static/gui/src/app/jujulib/urls.js [hunk 6]
```diff
@@ -351,7 +351,15 @@ var module = module;
   };
 
   const jujulib = exports.jujulib;
+
+  // Export patterns.
+  jujulib.NAME_PATTERN = NAME_PATTERN;
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/urls.js]: What about exporting just what we need (i.e. the test?). Something like:
```javascript
jujulib.isValidCharmName = namePattern.test.bind(namePattern);
...
```
?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8295/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8296/

**Comment by makyo** [comment]
:shipit:

**Pull request merged**